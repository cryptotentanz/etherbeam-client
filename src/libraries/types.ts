import { BigNumber } from 'ethers'

export const WETH_HASH = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'
export const USDC_HASH = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48'

export enum UserType {
  User = 'user',
  Administrator = 'administrator',
}

export enum AddressType {
  Unknown = 'unknown',
  Wallet = 'wallet',
  Token = 'token',
  Contract = 'contract',
  Transaction = 'transaction',
}

export enum TransactionMethodParameterType {
  Unknown = 'unknown',
  UnsignedInteger = 'unsigned_integer',
  Address = 'address',
  Addresses = 'addresses',
}

export enum TransactionStatus {
  Pending = 'pending',
  Mined = 'mined',
  Validated = 'validated',
  Cancelled = 'cancelled',
}

export enum TransactionActionType {
  Unknown = 'unknown',
  Approval = 'approval',
  Transfer = 'transfer',
  Swap = 'swap',
}

export enum TransactionActionDirection {
  Buy = 'buy',
  Sell = 'sell',
}

export interface User {
  name: string
  email: string
  type: UserType
}

export interface TransactionMethodParameter {
  name: string
  type: TransactionMethodParameterType
  rawType: string
  value: TransactionMethodValueType
}

export enum HistoryTime {
  Hour = 'h',
  Day = 'd',
  Week = 'w',
  Month = 'm',
  Year = 'y',
}

export interface TokenPriceHistory {
  time: HistoryTime
  price: BigNumber
  ratio: number
}

export interface Token extends Address {
  name: string
  description?: string
  symbol: string
  decimals: number
  chartPair?: string | null
  price?: BigNumber | null
  priceHistory: TokenPriceHistory[]
  website?: string | null
  whitepaper?: string | null
  github?: string | null
  linkedin?: string | null
  facebook?: string | null
  reddit?: string | null
  twitter?: string | null
  telegram?: string | null
  discord?: string | null
  actions?: TransactionAction[] | null
}

export interface Address {
  hash: string
  url: string
  type: AddressType
  label?: string
}

export interface Transaction {
  blockNumber?: number | null
  dateTime: Date
  address: Address
  status: TransactionStatus
  from?: Address | null
  to?: Address | null
  value?: BigNumber | null
  gas: {
    unitPrice?: BigNumber | null
    limit?: number | null
    used?: number | null
    ratio?: number | null
    price?: BigNumber | null
  }
  action?: TransactionMethod | null
  logs?: TransactionMethod[] | null
}

export interface TransactionMethod {
  contract: Address
  name: string
  parameters: TransactionMethodParameter[]
}

export type TransactionMethodValueType = string | BigNumber | Address | Address[] | null

export interface TransactionAction {
  index: number
  transaction: Transaction
  type: TransactionActionType
  direction?: TransactionActionDirection
  holder: Address
  fromAddress?: Address | Token | null
  fromAmount?: BigNumber | null
  toAddress?: Address | Token | null
  toAmount?: BigNumber | null
}
