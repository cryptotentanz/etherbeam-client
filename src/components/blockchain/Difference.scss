@use '../../styles/resources/fonts';
@use '../../styles/resources/sizes';

.my-difference {
  &-prefix {
    font-weight: bold;
    padding-right: sizes.$spacer-s;
  }

  &-amount {
    font-family: fonts.$monospace;
  }
}
