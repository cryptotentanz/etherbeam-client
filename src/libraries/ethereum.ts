import { Address, AddressType } from './types'

const ETHERSCAN_URL_BASE = 'https://etherscan.io/'

export type EtherscanUrlType = 'block' | 'tx' | 'token' | 'address'

export const getEtherscanUrl = (type: EtherscanUrlType, id: string | number): string => {
  return `${ETHERSCAN_URL_BASE}${type}/${id}`
}

export const createTransactionAddress = (hash: string): Address => {
  return {
    hash,
    type: AddressType.Transaction,
    url: getEtherscanUrl('tx', hash),
  }
}
