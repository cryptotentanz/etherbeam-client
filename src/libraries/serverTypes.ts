import {
  AddressType,
  TransactionActionType,
  TransactionMethodParameterType,
  TransactionStatus,
  UserType,
} from './types'

export interface UserData {
  email: string
  name: string
  user_type: UserType
}

export interface AddressData {
  address_hash: string
  address_type: AddressType
  label: string
}

export interface TransactionMethodParameterData {
  name: string
  parameter_type: TransactionMethodParameterType
  raw_type: string
  raw_value?: string
  decimal_value?: string
  addresses?: {
    address_hash: string
    address?: AddressData
  }[]
}

export interface TokenData extends AddressData {
  name: string
  symbol: string
  decimals: number
  chart_pair?: string
  description?: string
  price?: string
  price_1h?: string
  price_1h_ratio?: number
  price_1d?: string
  price_1d_ratio?: number
  price_1w?: string
  price_1w_ratio?: number
  price_1m?: string
  price_1m_ratio?: number
  price_1y?: string
  price_1y_ratio?: number
  website?: string
  whitepaper?: string
  github?: string
  linkedin?: string
  facebook?: string
  reddit?: string
  twitter?: string
  telegram?: string
  discord?: string
  from_transaction_actions?: TokenActionData[]
  to_transaction_actions?: TokenActionData[]
}

export interface TransactionData {
  transaction_hash: string
  status: TransactionStatus
  datetime: Date
  block_number?: number
  from_address_hash?: string
  from_address?: AddressData
  to_address_hash?: string
  to_address?: AddressData
  value?: string
  gas_unit_price?: string
  gas_limit?: number
  gas_used?: number
  gas_ratio?: number
  gas_fee?: string
  transaction_method_action?: TransactionMethodData
  transaction_method_logs?: TransactionMethodData[]
}

export interface TransactionMethodData {
  contract: AddressData
  name: string
  parameters: TransactionMethodParameterData[]
}

export interface TokenActionData {
  block_transaction: TransactionData
  index: number
  action_type: TransactionActionType
  holder_address_hash: string
  holder_address?: AddressData
  from_address_hash?: string
  from_address?: AddressData | TokenData
  from_amount?: string
  to_address_hash?: string
  to_address?: AddressData | TokenData
  to_amount?: string
}
