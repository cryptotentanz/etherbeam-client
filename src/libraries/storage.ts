import Cookies, { CookieSetOptions } from 'universal-cookie'
import { AuthToken } from '@totentech/api-client'

const cookies = new Cookies()

export const removeAuthToken = (): void => {
  const options: CookieSetOptions = {
    path: '/',
  }
  cookies.remove('tokenType', options)
  cookies.remove('client', options)
  cookies.remove('uid', options)
  cookies.remove('accessToken', options)
  cookies.remove('expiry', options)
}

export const saveAuthToken = (authToken: AuthToken | null): void => {
  if (!authToken) return removeAuthToken()

  const { client, uid, accessToken, tokenType, expiry } = authToken
  const options: CookieSetOptions = {
    path: '/',
    expires: new Date(parseInt(expiry) * 1000),
    // TODO: Add the secure option when HTTPS
    // secure: true,
    sameSite: 'strict',
  }

  cookies.set('tokenType', tokenType, options)
  cookies.set('client', client, options)
  cookies.set('uid', uid, options)
  cookies.set('accessToken', accessToken, options)
  cookies.set('expiry', expiry, options)
}

export const loadAuthToken = (): AuthToken | null => {
  const authToken = {
    tokenType: cookies.get('tokenType'),
    client: cookies.get('client'),
    uid: cookies.get('uid'),
    accessToken: cookies.get('accessToken'),
    expiry: cookies.get('expiry'),
  }
  return !Object.values(authToken).filter((v) => !v).length ? authToken : null
}
