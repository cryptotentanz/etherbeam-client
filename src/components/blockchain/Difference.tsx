import React, { FC } from 'react'
import numeral from 'numeral'
import { HistoryTime } from 'libraries/types'
import Badge from '../Badge'
import './Difference.scss'

type SizeProp = 's' | 'm' | 'l' | 'xl'

interface DifferenceProps {
  ratio: number
  time: HistoryTime
  size?: SizeProp
}

const Difference: FC<DifferenceProps> = ({ ratio, time, size }: DifferenceProps) => {
  const getColor = () => {
    if (ratio > 1.01) {
      return 'positive'
    } else if (ratio < 0.99) {
      return 'negative'
    } else {
      return 'light'
    }
  }

  return (
    <Badge color={getColor()} size={size}>
      <span className="my-difference-prefix">{time.toUpperCase()}</span>
      <span className="my-difference-amount">{numeral(ratio - 1).format('+0%')}</span>
    </Badge>
  )
}

export default Difference
