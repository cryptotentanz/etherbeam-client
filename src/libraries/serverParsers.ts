import { BigNumber } from 'ethers'
import orderBy from 'lodash/orderBy'
import {
  TransactionData,
  TokenData,
  AddressData,
  TokenActionData,
  TransactionMethodParameterData,
  TransactionMethodData,
  UserData,
} from './serverTypes'
import {
  Transaction,
  Token,
  HistoryTime,
  Address,
  AddressType,
  TransactionMethodValueType,
  TransactionAction,
  TransactionActionDirection,
  TransactionMethodParameter,
  TransactionMethodParameterType,
  TransactionMethod,
  User,
} from './types'
import { createTransactionAddress, getEtherscanUrl } from './ethereum'

const parseBigNumberFromServer = (value: string): BigNumber => BigNumber.from(value.split('.')[0])

export const parseUserData = ({ name, email, user_type }: UserData): User => {
  return {
    name,
    email,
    type: user_type,
  }
}

export const parseAddressData = ({ address_hash, address_type, label }: AddressData): Address => {
  return {
    hash: address_hash,
    type: address_type,
    label,
    url: getEtherscanUrl('address', address_hash),
  }
}

const parseHashData = (hash: string, address?: AddressData | TokenData): Address => {
  if (!address) {
    return {
      hash,
      type: AddressType.Unknown,
      url: getEtherscanUrl('address', hash),
    }
  }

  switch (address.address_type) {
    case AddressType.Token:
      return parseTokenData(address as TokenData)
    default:
      return parseAddressData(address)
  }
}

const parseTransactionMethodValueType = ({
  parameter_type,
  decimal_value,
  addresses,
}: TransactionMethodParameterData): TransactionMethodValueType => {
  switch (parameter_type) {
    case TransactionMethodParameterType.UnsignedInteger:
      return decimal_value ? parseBigNumberFromServer(decimal_value) : null
    case TransactionMethodParameterType.Address:
      return addresses && addresses?.length > 0 ? parseHashData(addresses[0].address_hash, addresses[0].address) : null
    case TransactionMethodParameterType.Addresses:
      return addresses ? addresses.map((address) => parseHashData(address.address_hash, address.address)) : null
    default:
      return null
  }
}

const parseTransactionMethodParameter = (parameter: TransactionMethodParameterData): TransactionMethodParameter => {
  const { name, parameter_type, raw_type } = parameter

  return {
    name,
    type: parameter_type,
    rawType: raw_type,
    value: parseTransactionMethodValueType(parameter),
  }
}

const parseTransactionMethodData = ({ contract, name, parameters }: TransactionMethodData): TransactionMethod => {
  return {
    contract: parseAddressData(contract),
    name,
    parameters: parameters.map((parameter) => parseTransactionMethodParameter(parameter)),
  }
}

export const parseTransactionData = ({
  transaction_hash,
  status,
  datetime,
  block_number,
  from_address_hash,
  from_address,
  to_address_hash,
  to_address,
  value,
  gas_unit_price,
  gas_limit,
  gas_used,
  gas_ratio,
  gas_fee,
  transaction_method_action,
  transaction_method_logs,
}: TransactionData): Transaction => {
  return {
    blockNumber: block_number,
    dateTime: datetime,
    address: createTransactionAddress(transaction_hash),
    status,
    from: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    to: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    value: value ? parseBigNumberFromServer(value) : null,
    gas: {
      unitPrice: gas_unit_price ? parseBigNumberFromServer(gas_unit_price) : null,
      limit: gas_limit,
      used: gas_used,
      ratio: gas_ratio,
      price: gas_fee ? parseBigNumberFromServer(gas_fee) : null,
    },
    action: transaction_method_action ? parseTransactionMethodData(transaction_method_action) : null,
    logs: transaction_method_logs?.map((log) => parseTransactionMethodData(log)),
  }
}

const parseTokenActionData = (
  {
    block_transaction,
    holder_address_hash,
    holder_address,
    index,
    action_type,
    from_address,
    from_address_hash,
    from_amount,
    to_address,
    to_address_hash,
    to_amount,
  }: TokenActionData,
  direction: TransactionActionDirection
): TransactionAction => {
  return {
    transaction: parseTransactionData(block_transaction),
    type: action_type,
    direction,
    index,
    holder: parseHashData(holder_address_hash, holder_address),
    fromAddress: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    fromAmount: from_amount ? parseBigNumberFromServer(from_amount) : null,
    toAddress: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    toAmount: to_amount ? parseBigNumberFromServer(to_amount) : null,
  }
}

export const parseTokenData = (data: TokenData): Token => {
  const {
    address_hash,
    name,
    description,
    symbol,
    decimals,
    chart_pair,
    price,
    price_1h,
    price_1h_ratio,
    price_1d,
    price_1d_ratio,
    price_1w,
    price_1w_ratio,
    price_1m,
    price_1m_ratio,
    price_1y,
    price_1y_ratio,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    from_transaction_actions,
    to_transaction_actions,
  } = data
  const priceHistory = []
  if (price_1h && price_1h_ratio) {
    priceHistory.push({
      time: HistoryTime.Hour,
      price: parseBigNumberFromServer(price_1h),
      ratio: price_1h_ratio as number,
    })
  }
  if (price_1d && price_1d_ratio) {
    priceHistory.push({
      time: HistoryTime.Day,
      price: parseBigNumberFromServer(price_1d),
      ratio: price_1d_ratio as number,
    })
  }
  if (price_1w && price_1w_ratio) {
    priceHistory.push({
      time: HistoryTime.Week,
      price: parseBigNumberFromServer(price_1w),
      ratio: price_1w_ratio as number,
    })
  }
  if (price_1m && price_1m_ratio) {
    priceHistory.push({
      time: HistoryTime.Month,
      price: parseBigNumberFromServer(price_1m),
      ratio: price_1m_ratio as number,
    })
  }
  if (price_1y && price_1y_ratio) {
    priceHistory.push({
      time: HistoryTime.Year,
      price: parseBigNumberFromServer(price_1y),
      ratio: price_1y_ratio as number,
    })
  }

  const fromActions = from_transaction_actions
    ? from_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Sell))
    : []
  const toActions = to_transaction_actions
    ? to_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Buy))
    : []
  const actions =
    from_transaction_actions || to_transaction_actions
      ? orderBy([...fromActions, ...toActions], ['transaction.dateTime', 'index'], ['desc', 'desc'])
      : null

  return {
    ...parseAddressData(data),
    type: AddressType.Token,
    url: getEtherscanUrl('token', address_hash),
    name,
    description,
    symbol,
    decimals,
    chartPair: chart_pair,
    price: price ? parseBigNumberFromServer(price) : null,
    priceHistory,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    actions,
  }
}
