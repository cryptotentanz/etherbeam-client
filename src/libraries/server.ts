import compact from 'lodash/compact'
import {
  HttpStatus,
  AuthToken,
  RestAction,
  callApi,
  ApiClient,
  createApiClient,
  setAuthToken,
  subscribeToNewTokens,
  flushAuthToken,
} from '@totentech/api-client'
import { Token, USDC_HASH, User, WETH_HASH } from './types'
import { TokenData, UserData } from './serverTypes'
import { parseTokenData, parseUserData } from './serverParsers'
import { loadAuthToken, removeAuthToken, saveAuthToken } from './storage'

const TIMEOUT = 60 * 1000

type UserUpdateCallback = (user: User | null) => void

export interface ServerData {
  apiClient: ApiClient
  userSubscriptions: UserUpdateCallback[]
}

const onNewAuthToken = (data: ServerData, token: AuthToken): void => {
  saveAuthToken(token)

  if (!token) {
    onUserUpdate(data, null)
  }
}

const initializeAuthToken = (apiClient: ApiClient): void => {
  const token = loadAuthToken()

  if (token) setAuthToken(apiClient, token)
}

export const initializeServerData = (): ServerData => {
  const onUserUpdateCallback = (user: User | null): void => {
    if (!user) removeAuthToken()
  }
  const data = {
    apiClient: createApiClient(process.env.SERVER_API_HOST as string, true, TIMEOUT),
    userSubscriptions: [onUserUpdateCallback],
  }

  initializeAuthToken(data.apiClient)
  subscribeToNewTokens(data.apiClient, (token) => onNewAuthToken(data, token))

  return data
}

export const subscribeToUserUpdates = ({ userSubscriptions }: ServerData, callback: UserUpdateCallback): void => {
  userSubscriptions.push(callback)
}

export const unsubscribeFromUserUpdates = ({ userSubscriptions }: ServerData, callback: UserUpdateCallback): void => {
  const index = userSubscriptions.indexOf(callback)

  if (index < 0) return

  userSubscriptions.splice(index, 1)
}

const onUserUpdate = ({ userSubscriptions }: ServerData, user: User | null): void => {
  userSubscriptions.forEach((callback) => callback(user))
}

export const signIn = async (serverData: ServerData, email: string, password: string): Promise<User | null> => {
  const payload = { email, password }

  const { data, status } = await callApi<UserData>(
    serverData.apiClient,
    RestAction.Post,
    'auth/sign_in',
    [HttpStatus.Ok],
    null,
    payload,
    true
  )

  const user = status != HttpStatus.Unauthorized ? parseUserData(data) : null
  onUserUpdate(serverData, user)

  return user
}

export const validateToken = async (serverData: ServerData): Promise<User | null> => {
  const { data, status } = await callApi<UserData>(
    serverData.apiClient,
    RestAction.Get,
    'auth/validate_token',
    [HttpStatus.Ok],
    null,
    null,
    true
  )

  const user = status != HttpStatus.Unauthorized ? parseUserData(data) : null
  onUserUpdate(serverData, user)

  return user
}

export const signOut = async (data: ServerData): Promise<void> => {
  const { apiClient } = data
  await callApi(apiClient, RestAction.Delete, 'auth/sign_out', [HttpStatus.Ok, HttpStatus.NotFound], null, null, true)

  flushAuthToken(apiClient)
  onUserUpdate(data, null)
}

export const fetchTokens = async (serverData: ServerData): Promise<Token[]> => {
  const { data } = await callApi<TokenData[]>(serverData.apiClient, RestAction.Get, 'contract_tokens')

  const ignoredHashes = [WETH_HASH, USDC_HASH]

  return compact(data.map((token) => (!ignoredHashes.includes(token.address_hash) ? parseTokenData(token) : null)))
}

export const fetchToken = async (
  serverData: ServerData,
  address: string,
  withList?: 'actions'
): Promise<Token | null> => {
  const { status, data } = await callApi<TokenData>(
    serverData.apiClient,
    RestAction.Get,
    `contract_tokens/${address}?list=${withList || ''}`,
    [HttpStatus.Ok, HttpStatus.NotFound]
  )

  if (status != HttpStatus.Ok) return null

  return parseTokenData(data)
}
