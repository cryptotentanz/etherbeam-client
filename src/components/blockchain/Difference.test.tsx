import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { HistoryTime } from 'libraries/types'
import Difference from './Difference'

describe('<Difference />', () => {
  let component: ShallowWrapper

  describe('Neutral', () => {
    beforeAll(() => {
      component = shallow(<Difference time={HistoryTime.Hour} ratio={1} />)
    })

    it('should render badge', () => expect(component.find('Badge').prop('color')).toBe('light'))

    it('should render prefix', () => {
      const span = component.find('Badge').find('span').at(0)
      expect(span.hasClass('my-difference-prefix')).toBeTruthy()
      expect(span.prop('children')).toBe('H')
    })

    it('should render amount', () => {
      const span = component.find('Badge').find('span').at(1)
      expect(span.hasClass('my-difference-amount')).toBeTruthy()
      expect(span.prop('children')).toBe('+0%')
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Positive', () => {
    beforeAll(() => {
      component = shallow(<Difference time={HistoryTime.Hour} ratio={1.75} />)
    })

    it('should render badge', () => expect(component.find('Badge').prop('color')).toBe('positive'))

    it('should render amount', () => expect(component.find('Badge').find('span').at(1).prop('children')).toBe('+75%'))
  })

  describe('Negative', () => {
    beforeAll(() => {
      component = shallow(<Difference time={HistoryTime.Hour} ratio={0.75} />)
    })

    it('should render badge', () => expect(component.find('Badge').prop('color')).toBe('negative'))

    it('should render amount', () => expect(component.find('Badge').find('span').at(1).prop('children')).toBe('-25%'))
  })

  describe('Time', () => {
    describe('Day', () => {
      beforeAll(() => {
        component = shallow(<Difference time={HistoryTime.Day} ratio={1} />)
      })

      it('should render prefix', () => expect(component.find('Badge').find('span').at(0).prop('children')).toBe('D'))
    })

    describe('Week', () => {
      beforeAll(() => {
        component = shallow(<Difference time={HistoryTime.Week} ratio={1} />)
      })

      it('should render prefix', () => expect(component.find('Badge').find('span').at(0).prop('children')).toBe('W'))
    })

    describe('Month', () => {
      beforeAll(() => {
        component = shallow(<Difference time={HistoryTime.Month} ratio={1} />)
      })

      it('should render prefix', () => expect(component.find('Badge').find('span').at(0).prop('children')).toBe('M'))
    })

    describe('Year', () => {
      beforeAll(() => {
        component = shallow(<Difference time={HistoryTime.Year} ratio={1} />)
      })

      it('should render prefix', () => expect(component.find('Badge').find('span').at(0).prop('children')).toBe('Y'))
    })
  })
})
