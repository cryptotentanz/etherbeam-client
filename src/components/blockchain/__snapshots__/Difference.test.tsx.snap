// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Difference /> Neutral should match snapshot 1`] = `
<span
  class="my-badge my-light-bg  "
>
  <span
    class="my-difference-prefix"
  >
    H
  </span>
  <span
    class="my-difference-amount"
  >
    +0%
  </span>
</span>
`;
